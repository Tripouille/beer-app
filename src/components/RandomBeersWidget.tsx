"use client";
import { styled, theme } from "@/styles/config";
import { useRandomBeerQuery } from "@/use-cases/useRandomBeerQuery";
import Image from "next/image";
import { useState } from "react";
import { nanoid } from "nanoid";
import Link from "next/link";
import { ROUTES } from "@/utils/routes";
import {
  leftSlideAnimation,
  rightSlideAnimation,
  twinkleAnimation,
} from "@/styles/animations";
import getBeerImageSrc from "@/domain/beersUtils";

export const RandomBeersWidget = () => {
  return (
    <RandomBeersWidgetContainer>
      <RandomBeerCard />
      <RandomBeerCard />
    </RandomBeersWidgetContainer>
  );
};

const RandomBeersWidgetContainer = styled("aside", {
  "--beer-card-min-width": "20em",
  "--beer-card-min-height": "12em",
  "--beer-card-border-radius": theme.radius.sm,
  display: "flex",
  flexWrap: "wrap",
  justifyContent: "center",
  gap: theme.spaces.md,
});

export const RandomBeerCard = () => {
  const [randomBeerID] = useState(nanoid);
  const {
    data: beer,
    isLoading,
    isError,
  } = useRandomBeerQuery(randomBeerID, { refetchInterval: 10000 });

  if (isLoading) return null;
  if (isError) return null;

  return (
    <BeerCard key={beer.id}>
      <Link href={ROUTES.BEER_DETAILS(beer.id)}>
        <Image alt={beer.name} src={getBeerImageSrc(beer)} fill sizes="2em" />
        <h2>{beer.name}</h2>
        <time>{beer.first_brewed}</time>
      </Link>
    </BeerCard>
  );
};

const BeerCard = styled("article", {
  overflow: "hidden",
  "&:hover": {
    transform: "scale(1.05)",
  },

  "@media (prefers-reduced-motion: no-preference)": {
    "&:nth-of-type(odd) a": {
      animation: `${rightSlideAnimation} 0.4s ease-out`,
    },
    "&:nth-of-type(even) a": {
      animation: `${leftSlideAnimation} 0.4s ease-out`,
    },
  },

  a: {
    height: "100%",
    display: "grid",
    gridTemplateAreas: `'image title'
                        'image first_brewed'`,
    gridTemplateColumns: "2em 1fr",
    placeItems: "center",
    width: "var(--beer-card-min-width)",
    minHeight: "var(--beer-card-min-height)",
    border: `2px solid ${theme.colors.identity500}`,
    borderRadius: "var(--beer-card-border-radius)",
    padding: theme.spaces.sm,
    gap: theme.spaces.sm,
    position: "relative",
    textDecoration: "none",
    transition: "transform 0.1s ease-in-out",

    "&:hover": {
      background: theme.colors.neutral100,
    },
  },

  "h2, time": {
    color: theme.colors.identity500,
  },

  img: {
    gridArea: "image",
    placeSelf: "center",
    objectFit: "contain",
  },

  h2: {
    gridArea: "title",
    justifySelf: "start",
    alignSelf: "end",
  },

  time: {
    gridArea: "first_brewed",
    placeSelf: "start",
  },
});

const BeerCardSkeleton = styled("div", {
  width: "var(--beer-card-min-width)",
  height: "var(--beer-card-min-height)",
  borderRadius: "var(--beer-card-border-radius)",
  background: theme.colors.neutral400,

  "@media (prefers-reduced-motion: no-preference)": {
    animation: `${twinkleAnimation} 1s linear infinite alternate`,
  },
});
