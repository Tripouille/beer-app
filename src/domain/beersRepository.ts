import { fetcher, FetcherSearchParams } from "@/utils/fetcher";
import { removeEmptyStringFromObject } from "@/utils/removeEmptyStringFromObject";
import { z } from "zod";

export const beerIDSchema = z.number().brand("beerID");
export type BeerID = z.infer<typeof beerIDSchema>;

export const beerSchema = z.object({
  id: beerIDSchema,
  name: z.string(),
  description: z.string(),
  first_brewed: z.string(),
  image_url: z.string().nullish(),
});
export type Beer = z.infer<typeof beerSchema>;

export const beersSchema = z.array(beerSchema);
export type Beers = z.infer<typeof beersSchema>;

export interface GetBeersParams extends FetcherSearchParams {
  per_page: string;
  beer_name: string;
}
export interface BeersRepository {
  getRandomBeer(): Promise<Beer>;
  getBeers(params: GetBeersParams): Promise<Beer[]>;
}

export const realBeersRepository: BeersRepository = {
  async getRandomBeer() {
    const response = await fetcher({
      url: "https://api.punkapi.com/v2/beers/random",
    });
    const data = (await response.json()) as Beer[];
    return beerSchema.parse(data[0]);
  },
  async getBeers(params) {
    const response = await fetcher({
      url: "https://api.punkapi.com/v2/beers",
      searchParams: removeEmptyStringFromObject(params),
    });
    const beers = (await response.json()) as Beer[];
    return beersSchema.parse(beers);
  },
};
