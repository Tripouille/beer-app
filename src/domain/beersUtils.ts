import { Beer } from "./beersRepository";

const IMG_SRC_FALLBACK = "https://images.punkapi.com/v2/keg.png";
export default function getBeerImageSrc(beer: Pick<Beer, "image_url">): string {
  return beer.image_url ?? IMG_SRC_FALLBACK;
}
