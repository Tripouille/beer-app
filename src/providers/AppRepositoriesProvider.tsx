"use client";

import { BeersRepository } from "@/domain/beersRepository";
import { invariant } from "@/utils/invariant";
import { ReactElement, createContext, useContext } from "react";

export interface AppRepositories {
  beersRepository: BeersRepository;
}

const AppRepositoriesContext = createContext<AppRepositories | null>(null);

export function AppRepositoriesProvider({
  appRepositories,
  children,
}: {
  appRepositories: AppRepositories;
  children: ReactElement;
}) {
  return (
    <AppRepositoriesContext.Provider value={appRepositories}>
      {children}
    </AppRepositoriesContext.Provider>
  );
}

export const useAppRepositories = (): AppRepositories => {
  const context = useContext(AppRepositoriesContext);

  invariant(
    context,
    "could not find AppRepositoriesContext, please ensure the component is wrapped in a <AppRepositoriesProvider>"
  );

  return context;
};
