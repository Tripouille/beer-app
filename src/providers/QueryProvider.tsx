"use client";

import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { ReactElement } from "react";
import {
  AppRepositoriesProvider,
  AppRepositories,
} from "./AppRepositoriesProvider";
import { realBeersRepository } from "@/domain/beersRepository";
import { ReactQueryDevtools } from "@tanstack/react-query-devtools";

const defaultQueryClient = new QueryClient({
  defaultOptions: {
    queries: {
      staleTime: 1000 * 60,
    },
  },
});

const defaultAppRepositories: AppRepositories = {
  beersRepository: realBeersRepository,
};

export function QueryProvider({
  children,
  queryClient,
  appRepositories,
}: {
  children: ReactElement;
  queryClient?: QueryClient;
  appRepositories?: AppRepositories;
}) {
  return (
    <AppRepositoriesProvider
      appRepositories={appRepositories ?? defaultAppRepositories}
    >
      <QueryClientProvider client={queryClient ?? defaultQueryClient}>
        {children}
        <ReactQueryDevtools initialIsOpen={false} />
      </QueryClientProvider>
    </AppRepositoriesProvider>
  );
}
