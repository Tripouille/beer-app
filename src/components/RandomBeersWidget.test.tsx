import { it, describe, vi, expect } from "vitest";
import { cleanup, render } from "@testing-library/react";
import { RandomBeersWidget } from "./RandomBeersWidget";
import { AppRepositories } from "@/providers/AppRepositoriesProvider";
import { QueryProvider } from "@/providers/QueryProvider";
import { QueryClient } from "@tanstack/react-query";
import { Beer, BeerID, beerSchema } from "@/domain/beersRepository";

type GetRamdomBeer = AppRepositories["beersRepository"]["getRandomBeer"];

function mount(getRandomBeer: GetRamdomBeer) {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });
  const fakeAppRepositories = {
    beersRepository: {
      getRandomBeer,
    },
  } as AppRepositories;

  cleanup();
  return render(
    <QueryProvider
      queryClient={queryClient}
      appRepositories={fakeAppRepositories}
    >
      <RandomBeersWidget />
    </QueryProvider>
  );
}

const FakeBeer: Beer = {
  id: 1 as BeerID,
  name: "Beer",
  description: "Description",
  image_url: "https://images.punkapi.com/v2/keg.png",
  first_brewed: "09/2007",
};

describe(RandomBeersWidget.name, () => {
  it("asks for 2 random beers", async () => {
    const getRandomBeer = vi.fn().mockImplementation(async () => FakeBeer);

    const component = mount(getRandomBeer);
    await component.findAllByTestId("loading-state");

    expect(getRandomBeer).toBeCalledTimes(2);
  });

  it("shows 2 random beers", async () => {
    let id = 0;
    const getRandomBeer = vi.fn().mockImplementation(async () => {
      id++;
      return beerSchema.parse({
        id,
        name: "name_" + id,
        description: "description_" + id,
        image_url: "/image_url_" + id,
        first_brewed: "first_brewed_" + id,
      });
    });

    const component = mount(getRandomBeer);
    await component.findByText("name_1");
    await component.findByText("name_2");

    expect(component.container).toMatchSnapshot();
  });

  it("shows 2 loading beers", async () => {
    const pendingPromise = new Promise(function () {});
    const getRandomBeer = vi.fn().mockReturnValue(pendingPromise);

    const component = mount(getRandomBeer);
    await component.findAllByTestId("loading-state");

    expect(component.container).toMatchSnapshot();
  });

  it("shows 2 error messages", async () => {
    const getRandomBeer = vi.fn().mockReturnValue(Promise.reject(""));

    const component = mount(getRandomBeer);
    await component.findAllByTestId("error-state");

    expect(component.container).toMatchSnapshot();
  });
});
