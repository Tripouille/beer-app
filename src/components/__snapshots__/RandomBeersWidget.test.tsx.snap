// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RandomBeersWidget > shows 2 error messages 1`] = `
<div>
  <aside
    class="c-hTSGOD"
  >
    <div
      class="c-bQpehp"
      data-testid="error-state"
    >
      <button
        class="c-eqmden"
      >
        Retry
      </button>
    </div>
    <div
      class="c-bQpehp"
      data-testid="error-state"
    >
      <button
        class="c-eqmden"
      >
        Retry
      </button>
    </div>
  </aside>
</div>
`;

exports[`RandomBeersWidget > shows 2 loading beers 1`] = `
<div>
  <aside
    class="c-hTSGOD"
  >
    <div
      aria-busy="true"
      class="c-catuKv"
      data-testid="loading-state"
    />
    <div
      aria-busy="true"
      class="c-catuKv"
      data-testid="loading-state"
    />
  </aside>
</div>
`;

exports[`RandomBeersWidget > shows 2 random beers 1`] = `
<div>
  <aside
    class="c-hTSGOD"
  >
    <article
      class="c-CVkVn"
    >
      <a
        href="/beers/1"
      >
        <img
          alt="name_1"
          src="/image_url_1"
        />
        <h2>
          name_1
        </h2>
        <time>
          first_brewed_1
        </time>
      </a>
    </article>
    <article
      class="c-CVkVn"
    >
      <a
        href="/beers/2"
      >
        <img
          alt="name_2"
          src="/image_url_2"
        />
        <h2>
          name_2
        </h2>
        <time>
          first_brewed_2
        </time>
      </a>
    </article>
  </aside>
</div>
`;
